src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -598,41 +598,41 @@ internal void Cancel(object caller)
                 // Keep looping until we either grabbed the lock (and therefore sent attention) or the connection closes\breaks
                 while ((!hasLock) && (_parser.State != TdsParserState.Closed) && (_parser.State != TdsParserState.Broken))
                 {
+                    // try to take the lock so that if another command is attempted it will queue on the lock
+                    // but don't require that the lock be taken because otherwise attention cannot be sent
+                    // during command execution causing cancellation to wait
+                    // This lock is also protecting against concurrent close and async continuations
                     Monitor.TryEnter(this, _waitForCancellationLockPollTimeout, ref hasLock);
-                    if (hasLock)
-                    { // Lock for the time being - since we need to synchronize the attention send.
-                      // This lock is also protecting against concurrent close and async continuations
 
-                        // Ensure that, once we have the lock, that we are still the owner
-                        if ((!_cancelled) && (_cancellationOwner.Target == caller))
-                        {
-                            _cancelled = true;
+                    // Ensure that that we are still the owner
+                    if ((!_cancelled) && (_cancellationOwner.Target == caller))
+                    {
+                        _cancelled = true;
 
-                            if (_pendingData && !_attentionSent)
+                        if (_pendingData && !_attentionSent)
+                        {
+                            bool hasParserLock = false;
+                            // Keep looping until we have the parser lock (and so are allowed to write), or the connection closes\breaks
+                            while ((!hasParserLock) && (_parser.State != TdsParserState.Closed) && (_parser.State != TdsParserState.Broken))
                             {
-                                bool hasParserLock = false;
-                                // Keep looping until we have the parser lock (and so are allowed to write), or the connection closes\breaks
-                                while ((!hasParserLock) && (_parser.State != TdsParserState.Closed) && (_parser.State != TdsParserState.Broken))
+                                try
                                 {
-                                    try
+                                    _parser.Connection._parserLock.Wait(canReleaseFromAnyThread: false, timeout: _waitForCancellationLockPollTimeout, lockTaken: ref hasParserLock);
```

Comment:
Feels like there should be an async version of this function. If the locks were async, wouldn't that also fix the cancellation hang.